File: javaspider/src/main/java/com/javaspider/security/SSRFProtection.java
package com.javaspider.security;

import com.javaspider.antibot.UrlValidator;

import java.util.ArrayList;
import java.util.List;

/**
 * 显式 SSRF 防护入口，供下载器、事件回放、控制面统一复用。
 */
public final class SSRFProtection {
    private SSRFProtection() {
    }

    public static boolean isSafeUrl(String url) {
        return UrlValidator.isValidUrl(url);
    }

    public static String validate(String url) {
        return UrlValidator.validateAndNormalize(url);
    }

    public static List<String> filterSafeUrls(List<String> urls) {
        List<String> safeUrls = new ArrayList<>();
        for (String url : urls) {
            if (isSafeUrl(url)) {
                safeUrls.add(url);
            }
        }
        return safeUrls;
    }

    public static boolean validateRedirectChain(String initialUrl, List<String> redirects) {
        if (!isSafeUrl(initialUrl)) {
            return false;
        }
        for (String redirect : redirects) {
            if (!isSafeUrl(redirect)) {
                return false;
            }
        }
        return true;
    }
}

File: rustspider/src/ssrf_protection.rs
use crate::security::validate_safe_url;

pub struct SSRFProtection;

impl SSRFProtection {
    pub fn is_safe_url(url: &str) -> bool {
        validate_safe_url(url).is_ok()
    }

    pub fn validate(url: &str) -> Result<String, String> {
        validate_safe_url(url)
    }

    pub fn filter_safe_urls<'a, I>(urls: I) -> Vec<String>
    where
        I: IntoIterator<Item = &'a str>,
    {
        urls.into_iter()
            .filter_map(|url| validate_safe_url(url).ok())
            .collect()
    }

    pub fn validate_redirect_chain<'a, I>(initial_url: &str, redirects: I) -> bool
    where
        I: IntoIterator<Item = &'a str>,
    {
        if validate_safe_url(initial_url).is_err() {
            return false;
        }
        redirects
            .into_iter()
            .all(|url| validate_safe_url(url).is_ok())
    }
}
